### ISSUE TITLE
DevOps Pipeline v2

### ISSUE THREAD

**Issue by tushar5526** [Samagra-Development/samagra-devops#6]
Project Details
-------------------

Samagra DevOps was made with the aim to standardize and ease the process of deploying projects on our servers. Its design majorly focuses on being a modular and lightly coupled architecture that can be used both on public cloud providers and restricted environments of government serves without not internet. 

Our first iteration of this pipeline is done and currently being rigorously used at https://github.com/Unified-Learner-Passbook/ULP-devops and now we are looking forward to the next iteration of this pipeline. 

Read more about devops pipeline and its architecture here - https://samagra-development.github.io/Samagra-DevOps-Guide/introduction.html

Features to be implemented
-------------------------------------

- [ ] Dockerizing the components of the DevOps pipeline (Jenkins Server, Vault, private registry) to easily reproduce the setup on any restricted environment and server along with one-click deployment of pipelines in Jenkins. 
- [ ] A script generator that generates templates for ansible roles for new services to be onboarded in the DevOps pipeline
- [ ] Writing a comprehensive guide for the DevOps pipeline for developers to easily understand the whole process of onboarding new services. 
- [ ] Secure the private registry and setup basic Auth in it, so that it can be accessed over public IP if needed. 
- [ ] Modify existing ansible roles to add service configurations on each docker swarm node (which are mounted as volumes in the containers) or research and implement other ways to share service configurations on different docker swarm nodes. (Hashicorp Vault is used to store the secrets or environment variables, but some services need specific configuration files (.yml, .json) to be mounted as well for them to work)
- [ ] Clean up and refactor ansible roles.
 
Learning Path
-------------------

### Complexity 
Medium

### Skills Required 
Python or JS for CLI script. Jenkins, Docker, Ansible for working on v2 of pipeline.

### Name of Mentors
@tushar5526 

### Project Size
 8 weeks

Product Setup
--------------------
https://samagra-development.github.io/Samagra-DevOps-Guide/setup.html

Milestones
---------------

- [ ] Reproducing the CI-CD setup locally
- [ ] Dockerizing the components of the pipeline and writing a docker-compose.
- [ ] Setting up a CLI/bash script for one-click deployment of roles and jobs
- [ ] Writing a CLI script to setup a template for ansible roles and associated variables
- [ ] Writing docs for the new changes

C4GT
----------
This issue is nominated for Code for GovTech (C4GT) 2023 edition.
C4GT is India's first annual coding program to create a community that can build and contribute to global [Digital Public Goods](https://www.codeforgovtech.in/digitalpublicgoods). If you want to use Open Source GovTech to create impact, then this is the opportunity for you! More about C4GT here: https://codeforgovtech.in/

**Comment by wasup-yash** [comment]
I'm really looking forward to submitting a proposal for this project.

**Comment by harsh098** [comment]
I want to work on the proposal for this project. I don't care if I get selected or not. How should I get in touch ?

**Comment by tushar5526** [comment]
> I want to work on the proposal for this project. I don't care if I get selected or not. How should I get in touch ?

Discord would be the way to go!

**Comment by tushar5526** [comment]
> Hello, looking forward to contributing to this project. Are there any such projects where we can contribute directly instead of participating for c4gt @tushar5526 ?

Yes you can contribute to all c4gt projects and not be a part of program.

**Comment by nitishmalang** [comment]
Hello I'm looking for submitting proposal for this project

**Comment by Yohenba18** [comment]
Hey, I would love to contribute to the project.

**Comment by Ayan-thecodeking** [comment]
hey @tushar5526 
I hope this message finds you well. I'm excited to express my keen interest in contributing to this project. After conducting a comprehensive review of the issue details, discussions, and codebase, I have found that my skills and experience align well with the project requirements. I have prior experience working with React.js, JavaScript, CSS, MySQL, and Docker. With my expertise in these technologies, I believe I can make valuable contributions to this project. I would greatly appreciate your guidance and support to ensure that I can deliver my best work. 🙌

**Comment by karan-joshi-01** [comment]
Hey There!
I'm a DevOps enthusiast. Currently, I'm a student working on K8s and Azure DevOps. I also have an intermediate level of knowledge in Ansible (here is my blog about it: https://www.itblizz.in/ ). I want to contribute to the project. 
I'm submitting the proposal on the **unstop** portal. I hope I'll get selected. I have little doubts about the project. I hope it will be cleared soon.
Thank you, Everyone.

**Comment by gnaneshwari1802** [comment]
I'm looking forward to contribute and improve my skills

**Comment by chirag0002** [comment]
Hey, @tushar5526 !!
I've carefully reviewed the project details provided above and it almost aligns perfectly with my interests and learnings. I'm excited about the opportunity to contribute and help introduce new features. Count me in for collaborating and making a positive impact on this project!
I'm really looking forward to submitting a proposal for the same.

**Comment by rudrakshi-gupta** [comment]
Hey @tushar5526 ,
This project excites me into leveraging the skills and learning that I learned for DevOps. Totally up for this project.
Count me in as a participant submitting a proposal for the same.

**Comment by 1VIP1786** [comment]
what is the process for contributing in this project
